8338894: Deprecate jhsdb debugd for removal

kevinjwalls opened the pull request:
Deprecation annotations and warnings on starting the tool(s).
Handle man page in a separate issue.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8339458](https://bugs.openjdk.org/browse/JDK-8339458) to be approved

### Issues
 * [JDK-8338894](https://bugs.openjdk.org/browse/JDK-8338894): Deprecate jhsdb debugd for removal (**Enhancement** - P4)
 * [JDK-8339458](https://bugs.openjdk.org/browse/JDK-8339458): Deprecate jhsdb debugd for removal (**CSR**)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) Review applies to [73c0f49f](https://git.openjdk.org/jdk/pull/20830/files/73c0f49f1b69603d0c5f36099c42081c20f2081a)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20830/head:pull/20830` \
`$ git checkout pull/20830`

Update a local copy of the PR: \
`$ git checkout pull/20830` \
`$ git pull https://git.openjdk.org/jdk.git pull/20830/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20830`

View PR using the GUI difftool: \
`$ git pr show -t 20830`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20830.diff">https://git.openjdk.org/jdk/pull/20830.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20830#issuecomment-2325942155)

kevinjwalls commented:
src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HotSpotAgent.java:
Add annotation Deprecated(since="24", forRemoval=true) on the various startServer methods, and shutdownServer

src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:
Add deprecation warnings in help messages.

src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/DebugServer.java:
A standalone launcher, providing another main entry point (separate to jhsdb), which creates a HotSpotAgent() and calls startServer().  This class should have the same deprecation annotation, and print a warning in its main method.

Handle man page in a separate issue.

```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb
    clhsdb              command line debugger
    hsdb                ui debugger                                                                                                
    debugd --help       to get more information (deprecated)
    jstack --help       to get more information
    jmap   --help       to get more information
    jinfo  --help       to get more information
    jsnap  --help       to get more information
bash-4.2$      

bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb debugd --help                                                                   
WARNING: debugd is deprecated and will be removed in a future release.                                                           
  --serverid <id>     ...etc...
  
  
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb jstack --help
    --locks                 To print java.util.concurrent locks.                                                                 
    --mixed                 To print both Java and native frames (mixed mode).
    --pid <pid>             To attach to and operate on the given live process.
    --core <corefile>       To operate on the given core file.
    --exe <executable for corefile>
    --connect [<serverid>@]<host>[:registryport][/servername] To connect to a remote debug server (debugd).

WARNING: debugd (and therefore --connect) are deprecated and will be removed in a future release.

    The --core and --exe options must be set together to give the core..etc...


bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb debugd --pid 123                                                                
WARNING: debugd is deprecated and will be removed in a future release.                                                         
Attaching to process ID 123 and starting RMI services, please wait...  
...etc...
```

kevinjwalls commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HotSpotAgent.java:
```diff
@@ -204,6 +204,7 @@ public synchronized boolean detach() throws DebuggerException {
       specific debuggee on the server. Allows to specify the port number
       to which the RMI connector is bound. If not specified a random
       available port is used. */
+    @Deprecated(since="24", forRemoval=true)
```
Maybe we can call these tags pointless.  But they do give a useful check that our own code (which should be limited to the tool being deprecated here) has to be aware of and suppress the deprecation.

Also if anybody does have e.g. class MyTool extends DebugServer and was building with something like:
```
bash-4.2$ javac --add-modules jdk.hotspot.agent --add-exports jdk.hotspot.agent/sun.jvm.hotspot=ALL-UNNAMED MyTool.java        
```
..then that was building with no error, but now gets a Warning:
```
bash-4.2$ build/linux-x64/images/jdk/bin/javac --add-modules jdk.hotspot.agent --add-exports jdk.hotspot.agent/sun.jvm.hotspot=ALL-UNNAMED MyTool.java                                                                                                        
MyTool.java:3: warning: [removal] DebugServer in sun.jvm.hotspot has been deprecated and marked for removal
public class MyTool extends DebugServer {
                            ^
1 warning
```

..so they do offer some safeguard.

AlanBateman replied:
> Maybe we can call these tags pointless. But they do give a useful check that our own code (which should be limited to the tool being deprecated here) has to be aware of and suppress the deprecation.

If you find them useful then okay, main thing is that this is all internal to the jdk.hotspot.agent module. In some future release then this code will be removed and maybe the deprecated annotation will help identify the dead code.

kevinjwalls replied:
Yes all internal and the tags do not increase our commitment, just may help with finding edge cases.  I think they are useful and cheap, thanks.

AlanBateman replied:
It's okay to include them in the PR but don't list them in the CSR as they aren't part of any interface, they are just internal to this module.

AlanBateman reviewed (approved):
I can look at the CSR when it is ready.

kevinjwalls commented:
OK one other addition, as jhsdb debugd and the --connect option are two sides of the same feature:

We generally expect attaches between the same JDK version (see the SA version mismatch warning),
so a later JDK with the ability to connect to a previous JDK which still has debugd makes no sense.

We can warn at the tool running with --connect, e.g.

bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb jstack --connect serverid@debugserver:1234/servername
WARNING: debugd (and therefore --connect) are deprecated and will be removed in a future release.
Attaching to remote server serverid@debugserver:1234/servername, please wait...

In our own tests, the debugd warning message causes no problems, but the warning on using --connect causes some test disruption, will update.

AlanBateman commented:
> OK one other addition, as jhsdb debugd and the --connect option are two sides of the same feature:
> 
> We generally expect attaches between the same JDK version (see the SA version mismatch warning), so a later JDK with the ability to connect to a previous JDK which still has debugd makes no sense.
> 
> We can warn at the tool running with --connect, e.g.
> 
> ```
> bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb jstack --connect serverid@debugserver:1234/servername
> WARNING: debugd (and therefore --connect) are deprecated and will be removed in a future release.
> Attaching to remote server serverid@debugserver:1234/servername, please wait...
> ```

I agree it doesn't make sense. For the warning message it might be clear to put the tool name into message, as in something like "WARNING: jhsdb jstack --connect is deprecated ...").

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/DebugServer.java:
```diff
@@ -28,6 +28,7 @@
 import sun.jvm.hotspot.runtime.*;
 import sun.jvm.hotspot.oops.*;
 
+@Deprecated(since="24", forRemoval=true)
```
I didn't even realize this class existed. So this an (undocumented) way to start the debug server. Is there some other class that connects to it other than using `jhsdb --connect`?

kevinjwalls replied:
Right, I think this class is an orphan, it has no users.
The HotSpotAgent constructor mentions DebugServer in one comment, but does not use it, and probably means to say "startServer".  This comment should be updated when startServer is eventually removed.

Checking older releases, there used to be shell scripts which invoked DebugServer, and an SADebugServer class which invoked DebugServer.

AlanBateman commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:
```diff
@@ -69,6 +69,8 @@ private static boolean commonHelp(String mode, boolean canConnectToRemote) {
         if (canConnectToRemote) {
             System.out.println("    --connect [<serverid>@]<host>[:registryport][/servername] To connect to a remote debug server (debugd).");
+            System.out.println();
+            System.out.println("WARNING: debugd (and therefore --connect) are deprecated and will be removed in a future release.");
```
This is -help output so I assume should be "The --connect option is deprecated and will be removed in a future release".

AlanBateman reviewed (changes requested):
Can you paste in up to date examples of the output from `jhsdb debugd ..` and `jhsdb jstack --connect .." so we can see the warnings.  The output is in the first command has a few issues so would be good to see the latest.

kevinjwalls commented:
Updated, including tests as the new --connect warning needs to be ignored.
We were using OutputAnalyzer's stderrShouldBeEmptyIgnoreDeprecatedWarnings() but it's specifically ignoring ".* VM warning:.* deprecated.*".  These changes are easy to find and revert when removal happens.

Alan you mention "warnings in the --help output", I still have a warning in the middle of the help text here, I tried it elsewhere but liked having the warning about --connect next to where --connect is mentioned in the help.

It seems like unnecessary additional changes would be needed to enable the common argument helper method to know what command is being invoked to make the --connect message specific to jhsdb jstack --connect etc.  Also one message to check for in tests.  


Updated messages:

```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb
    clhsdb              command line debugger
    hsdb                ui debugger
    debugd --help       to get more information (deprecated)
    jstack --help       to get more information
    jmap   --help       to get more information
    jinfo  --help       to get more information
    jsnap  --help       to get more information
```

```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb debugd --help
WARNING: debugd is deprecated and will be removed in a future release.
    --serverid <id>         A unique identifier....
```	

	
```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb debugd --pid 123
WARNING: debugd is deprecated and will be removed in a future release.
Attaching to process ID 123 and starting RMI services, please wait...
```



```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb jstack --help
    --locks                 To print java.util.concurrent locks.
    --mixed                 To print both Java and native frames (mixed mode).
    --pid <pid>             To attach to and operate on the given live process.
    --core <corefile>       To operate on the given core file.
    --exe <executable for corefile>
    --connect [<serverid>@]<host>[:registryport][/servername] To connect to a remote debug server (debugd).

WARNING: --connect is deprecated and will be removed in a future release.

    The --core and --exe options...
```	
	
```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb jstack --connect serverid@debugserver:1234/servername
WARNING: --connect is deprecated and will be removed in a future release.
Attaching to remote server serverid@debugserver:1234/servername, please wait...
```

AlanBateman commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:
```diff
@@ -69,6 +69,8 @@ private static boolean commonHelp(String mode, boolean canConnectToRemote) {
         if (canConnectToRemote) {
             System.out.println("    --connect [<serverid>@]<host>[:registryport][/servername] To connect to a remote debug server (debugd).");
+            System.out.println();
+            System.err.println("WARNING: --connect is deprecated and will be removed in a future release.");
```
This is `jhsdb $TOOL --help` output. Shouldn't this just have "(deprecated)" before "To connect ..." rather than a blank line and warning?  I assume the changes should only print "WARNING" if the user attempts to run `jhsdb $TOOL -connect ..`.

kevinjwalls replied:
OK yes understood, we can keep things less ALARMING unless you try and use the options.

--connect is already a long line, so adding "(deprecated)" I think gets lost.  We can do:

```
bash-4.2$ build/linux-x64/images/jdk/bin/jhsdb jstack --help
    --locks                 To print java.util.concurrent locks.
    --mixed                 To print both Java and native frames (mixed mode).
    --pid <pid>             To attach to and operate on the given live process.
    --core <corefile>       To operate on the given core file.
    --exe <executable for corefile>
    --connect [<serverid>@]<host>[:registryport][/servername] To connect to a remote debug server (debugd).

    The --connect is deprecated and will be removed in a future release.
    The --core and --exe options must be set together to give the core
    file, and associated executable, to operate on. They can use
    absolute or relative paths.
    The --pid option can be set to operate on a live process.
    The --connect option can be set to connect to a debug server (debugd).
    --core, --pid, and --connect are mutually exclusive.

    Examples: jhsdb jstack --pid 1234
          or  jhsdb jstack --core ./core.1234 --exe ./myexe
          or  jhsdb jstack --connect serverid@debugserver:1234/servername
bash-4.2$
```


Actually the text "The --connect option can be set to connect to a debug server (debugd)." is in there twice.  Will remove the second one.

AlanBateman commented:
> you mention "warnings in the --help output", I still have a warning in the middle of the help text here, I tried it elsewhere but liked having the warning about --connect next to where --connect is mentioned in the help.

I had added a comment before seeing your latest update. My view is that it would be better to only print "WARNING" when the user tried to use a deprecated command/option. In the help output it can show that the command/option is deprecated but blank line + WARNING in the help looks a bit unusual (to me anyway).

AlanBateman reviewed (approved):
The WARNING and help output look okay now.

kevinjwalls commented:
Thanks, have updated CSR also.

plummercj commented on src/jdk.hotspot.agent/doc/index.html:
```diff
@@ -21,9 +21,11 @@ <h3>SA Modes</h3>
 </ul>
 <p>
+<p><strong>WARNING: <b>jhsdb debugd</b> is deprecated and will be removed in a future release.</strong></p>
+<p>
```
You have an extra `<p>` here. I think the one on line 25 should be removed.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:
```diff
@@ -71,12 +71,14 @@ private static boolean commonHelp(String mode, boolean canConnectToRemote) {
         }
         System.out.println();
+        if (canConnectToRemote) {
+            System.out.println("    The --connect option is deprecated and will be removed in a future release.");
```
It is ok to add this, but now there is no longer any help message to explain what `--connect` does.

kevinjwalls replied:
We have line 71 saying "--connect .....To connect to a remote debug server (debugd)."
It's a long line and scanning it people might not read it all, but too late to worry overly about that. 8-)

plummercj replied:
Ok. I was looking at the deletion of line 79. I didn't realize it was removed because it pretty much just repeated line 71.

plummercj commented on test/hotspot/jtreg/serviceability/sa/sadebugd/ClhsdbTestConnectArgument.java:
```diff
@@ -75,7 +75,10 @@ public static void main(String[] args) throws Exception {
 
-            out.stderrShouldBeEmptyIgnoreDeprecatedWarnings();
+            out.shouldMatch("WARNING: --connect is deprecated");
+            // --connect is deprecated.  When removed, can revert to:
+            // out.stderrShouldBeEmptyIgnoreDeprecatedWarnings();
```
I don't think this comment is necessary since this test is going way once `--connect` is removed.  Same with this comment in all the other tests.

plummercj commented on test/hotspot/jtreg/serviceability/sa/sadebugd/ClhsdbTestConnectArgument.java:
```diff
@@ -75,7 +75,10 @@ public static void main(String[] args) throws Exception {
             System.err.println(out.getStderr());
 
-            out.stderrShouldBeEmptyIgnoreDeprecatedWarnings();
+            out.shouldMatch("WARNING: --connect is deprecated");
```
You might want to consider improving stderrShouldBeEmptyIgnoreDeprecatedWarnings(). It could, for instance, do a case insensitive search for "warning" and "deprecated" rather than the more specific search it is currently doing. Just a suggestion. Don't bother if you don't think it is worth it.

kevinjwalls replied:
I had looked at doing that but was worried about missing other real messages.
However looking again, to match "^WARNING: .* is deprecated.*" seems quite unambiguous!
Will update...

plummercj reviewed (approved):
Looks good. Make sure you get good test coverage of you OutputAnalyzer change.

kevinjwalls commented:
> Make sure you get good test coverage of you OutputAnalyzer change.

That was the right thing to say, as I didn't have it quite right.  stderrShouldBeEmptyIgnoreDeprecatedWarnings() needs to work if both kinds of deprecation message occurred. 8-)
It needs to remove both patterns, then check, like in the other relevant method, stderrShouldMatchIgnoreDeprecatedWarnings().  This can be in a common method. Updated.

kevinjwalls commented:
/integrate
